nav2_loopback_sim/src/loopback_simulator.cpp

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2023 Smit Dumore
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "nav2_loopback_sim/loopback_simulator.hpp"
+
+namespace nav2_loopback_sim
+{
+
+LoopbackSimulator::LoopbackSimulator(const rclcpp::NodeOptions & options)
+: Node("loopback_simulator_node", options)
+{
+  // Initialize subscribers
+  vel_subscriber_ = this->create_subscription<geometry_msgs::msg::Twist>(
+    "cmd_vel", 10, std::bind(&LoopbackSimulator::twistCallback, this, std::placeholders::_1));
+
+  init_pose_subscriber_ = create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    "initialpose", 10,
+    std::bind(&LoopbackSimulator::initPoseCallback, this, std::placeholders::_1));
+
+  // Initialize publishers
+  scan_publisher_ = this->create_publisher<sensor_msgs::msg::LaserScan>("scan", 10);
+
+  // Frequency parameter for timer function
+  double timer_frequency{0.0};
+  declare_parameter("timer_frequency", 10.0);
```

Comment:
Question: the `dt` should that be run at the same controller rate? What happens if this is set to 0.1, but the controls are published every 0.05s (20hz)?

Reply:
Yes, ideally, when this simulator node is launched, it should get the 'dt' from the controller node. Because if the controller is running at a faster rate, then the simulator is missing out on control commands, which can lead to inaccuracies over time.